File: backend/worker.mjs
import { setTimeout as delay } from 'node:timers/promises';
import { openStore, isMain, now, audit, lockSettings } from './database.mjs';
import { replay } from './simulator.mjs';

export async function recover(store) {
  await store.transaction(async q => {
    await lockSettings(q, store);
    await q("UPDATE jobs SET status='queued' WHERE status='running'");
    await q("UPDATE strategies SET status='queued' WHERE status='running'");
  });
}
export async function processOne(store) {
  const work = await store.transaction(async q => {
    if ((await lockSettings(q, store)).halted) return null;
    const [job] = await q("SELECT * FROM jobs WHERE status='queued' ORDER BY created_at LIMIT 1");
    if (!job) return null;
    await q("UPDATE jobs SET status='running',updated_at=$1 WHERE id=$2", [now(), job.id]);
    await q("UPDATE strategies SET status='running' WHERE id=$1", [job.strategy_id]);
    const [strategy] = await q('SELECT * FROM strategies WHERE id=$1', [job.strategy_id]);
    return { job, strategy };
  });
  if (!work) return false;
  try {
    const { job, strategy: s } = work;
    const result = replay(s.symbol, s.capital, s.fast, s.slow);
    await store.transaction(async q => {
      const settings = await lockSettings(q, store);
      const [current] = await q('SELECT status FROM jobs WHERE id=$1', [job.id]);
      if (settings.halted || current.status !== 'running') return;
      await q("UPDATE jobs SET status='completed',result=$1,updated_at=$2 WHERE id=$3", [JSON.stringify(result), now(), job.id]);
      await q("UPDATE strategies SET status='ready',pnl=$1 WHERE id=$2", [result.pnl, s.id]);
      await audit(q, `Completed sample-data replay: ${s.name}. ${result.trades.length} simulated fills.`);
    });
  } catch {
    await store.transaction(async q => {
      await lockSettings(q, store);
      const [job] = await q('SELECT status FROM jobs WHERE id=$1', [work.job.id]);
      if (job.status !== 'running') return;
      await q("UPDATE jobs SET status='failed' WHERE id=$1", [work.job.id]);
      await q("UPDATE strategies SET status='failed' WHERE id=$1", [work.strategy.id]);
      await audit(q, 'Paper replay failed. Check worker logs and retry.');
    });
    console.error('Paper replay failed.');
  }
  return true;
}
if (isMain(import.meta.url)) {
  const store = openStore();
  let stopping = false;
  for (const signal of ['SIGINT', 'SIGTERM']) process.on(signal, () => { stopping = true; });
  try {
    await recover(store);
    console.log('Node.js paper replay worker ready. Run exactly one worker.');
    while (!stopping) if (!await processOne(store)) await delay(1000);
  } finally { await store.close(); }
}

File: backend/worker.py
"""Run one worker with python -m backend.worker. Database-backed paper replay queue."""
import json
import time
from sqlalchemy import select, update
from .database import Session, Job, Strategy, Settings, audit, initialize, now
from .simulator import replay


def process_one():
    with Session.begin() as db:
        if db.get(Settings, 1).halted:
            return False
        job = db.scalars(select(Job).where(Job.status == 'queued').order_by(Job.created_at)).first()
        if not job:
            return False
        claimed = db.execute(update(Job).where(Job.id == job.id, Job.status == 'queued').values(status='running', updated_at=now()))
        if claimed.rowcount != 1:
            return False
        strategy = db.get(Strategy, job.strategy_id)
        strategy.status = 'running'
        identifier, sid = job.id, strategy.id
        args = (strategy.symbol, strategy.capital, strategy.fast, strategy.slow)
    try:
        result = replay(*args)
        with Session.begin() as db:
            # Lock the same control row as pause-all before committing a result.
            settings = db.execute(select(Settings).where(Settings.id == 1).with_for_update()).scalar_one()
            job = db.get(Job, identifier)
            if settings.halted or job.status != 'running':
                return True
            job.status, job.result, job.updated_at = 'completed', json.dumps(result), now()
            strategy = db.get(Strategy, sid)
            strategy.status, strategy.pnl = 'ready', result['pnl']
            audit(db, f'Completed sample-data replay: {strategy.name}. {len(result["trades"])} simulated fills.')
    except Exception:
        with Session.begin() as db:
            job = db.get(Job, identifier)
            if job.status == 'running':
                job.status = 'failed'
                db.get(Strategy, sid).status = 'failed'
                audit(db, 'Paper replay failed. Check worker logs and retry.')
        import logging
        logging.exception('Paper replay failed')
    return True


if __name__ == '__main__':
    initialize()
    # This MVP supports a single worker. Requeue unfinished work after a restart.
    with Session.begin() as db:
        db.execute(update(Job).where(Job.status == 'running').values(status='queued'))
        db.execute(update(Strategy).where(Strategy.status == 'running').values(status='queued'))
    print('Paper replay worker ready.', flush=True)
    while True:
        if not process_one():
            time.sleep(1)
